### PR TITLE
Implement Subset Sum algorithm using Dynamic Programming.

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Description

In this PR, I use dynamic programming approach to implement Subset Sum problem.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

File: src/dynamic_programming/subset_sum.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        (false, None)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_subset_sum_exists() {
```
Comment (realstealthninja) [inline src/dynamic_programming/subset_sum.rs:55]: since nums is i32 could you add a test for negative numbers as well it negative numbers arent covered please use the unsigned equivalent

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: src/dynamic_programming/subset_sum.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:55]: Well, I forgot that this algorithm requires non-negative integers as input. I have updated the function signature that use unsigned integer type `u32` in Rust.

**Comment by Miraksi** [comment]
Is there a reason for your function to return `(bool, Option<Vec<u32>>)` and not just `Option<u32>`?
Since you only ever return `(true, Some(x))` and `(false, None)` I see no point in also returning a bool.

**Comment by TruongNhanNguyen** [comment]
> Is there a reason for your function to return `(bool, Option<Vec<u32>>)` and not just `Option<u32>`? Since you only ever return `(true, Some(x))` and `(false, None)` I see no point in also returning a bool.

The function will return a `bool` value to specify that exist subset (or non-exist). If bool value is true, we can return subset of the array that has sum of elements equal to given `target`, else return `None`.

**Review by realstealthninja — APPROVED** [review]
LGTM @vil02  could you have a look please.

File: src/dynamic_programming/subset_sum.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Fill the dp table
+    for (i, num) in nums.iter().enumerate() {
+        for j in 1..=target_sum as usize {
+            dp[i + 1][j] = dp[i][j];
+            if *num <= j as u32 {
+                dp[i + 1][j] |= dp[i][j - *num as usize];
+            }
+        }
+    }
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:26]: Please move it into a separate function.

File: src/dynamic_programming/subset_sum.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    let mut subset = Vec::new();
+    let mut i = n;
+    let mut j = target_sum as usize;
+    while i > 0 && j > 0 {
+        if dp[i][j] && !dp[i - 1][j] {
+            subset.push(nums[i - 1]);
+            j -= nums[i - 1] as usize;
+        }
+        i -= 1;
+    }
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:38]: Please move is into a separate function.

File: src/dynamic_programming/subset_sum.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        assert_eq!(subset, None);
+    }
+
+    #[test]
+    fn test_subset_sum_single_element_array() {
+        let nums = vec![7];
+        let target_sum = 7;
+        let subset = subset_sum(&nums, target_sum);
+        assert_eq!(subset, Some(vec![7]));
+    }
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:84]:
```suggestion
    macro_rules! test_subset_sum {
        ($($name:ident: $inputs:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (nums, target_sum, expected) = $inputs;
                    assert_eq!(subset_sum(&nums, target_sum), expected);
                }
            )*
        }
    }

    test_subset_sum! {
        subset_sum_exists: (vec![3, 34, 4, 12, 5, 2], 9, Some(vec![4, 5])),
        sum_does_not_exist: (vec![3, 34, 4, 12, 5, 2], 30, None),
        empty_array: (vec![], 5, None),
        length_1_array: (vec![7], 7, Some(vec![7])),
        length_1_array_no_solution: (vec![7], 8, None),
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/dynamic_programming/subset_sum.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
+/// Determines whether there exists a subset of the given array `nums` that adds up to the specified `target_sum`.
+/// If such a subset exists, returns the subset that adds up to `target_sum`, sorted in ascending order.
+/// If no such subset exists, returns `None`.
+///
+/// # Arguments
+///
+/// * `nums` - A slice of integers representing the input array.
+/// * `target_sum` - The target sum that the subset needs to achieve.
+pub fn subset_sum(nums: &[u32], target_sum: u32) -> Option<Vec<u32>> {
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:9]: - To identifies and handles duplicate elements correctly, when updating subsets in the `dp` table, we ensures that each subset is unique by using a clone operation to create new subsets. This ensures that duplicate elements do not affect the generation of subsets.
- Non-unique solutions refer to different combinations of elements that yield the same sum. We can generate all possible subsets that sum up to the target sum, including those with duplicate elements, and then removes duplicate subsets before returning the result. This ensures that the result contains only distinct subsets, each representing a unique combination of elements.

File: src/dynamic_programming/subset_sum.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    let n = nums.len();
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:10]: I have refactored the code base, so this unneeded `n` in no longer used.

File: src/dynamic_programming/subset_sum.rs (hunk 3 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:38]: Done

File: src/dynamic_programming/subset_sum.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs]: You can see updates on the latest commit.

File: src/dynamic_programming/subset_sum.rs (hunk 4 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs]: Done.

File: src/dynamic_programming/subset_sum.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+/// * `target_sum` - The target sum to which subsets are sought.
+///
+/// # Returns
+///
+/// An option containing a vector of vectors representing subsets that sum up to the target sum,
+/// or `None` if no such subsets exist.
+pub fn subset_sum(nums: &[u32], target_sum: u32) -> Option<Vec<Vec<u32>>> {
+    let mut dp = initialize_dp_table(nums.len(), target_sum);
+
+    fill_dp_table(nums, &mut dp);
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:15]: I think should look somehow like this:
```suggestion
    let dp = create_dp(nums, target_sum);
```
Furthermore, `dp` is not really a good name.

File: src/dynamic_programming/subset_sum.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        }
+    }
+
+    test_subset_sum! {
+        test_subset_sum_basic: (vec![1, 2, 3, 4, 5], 5, Some(vec![vec![1, 4], vec![2, 3], vec![5]])),
+        test_subset_sum_empty_input: (vec![], 10, None),
+        test_subset_sum_no_solution: (vec![1, 2, 3, 4, 5], 100, None),
+        test_subset_sum_single_element: (vec![7], 7, Some(vec![vec![7]])),
+        test_subset_sum_with_duplicates: (vec![1, 2, 2, 3], 5, Some(vec![vec![1, 2, 2], vec![2, 3]])),
+        test_subset_sum_with_non_unique_elements: (vec![1, 2, 3, 3, 4], 7, Some(vec![vec![1, 2, 4], vec![1, 3, 3], vec![3, 4]])),
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:132]: Do you somehow relay on the fact that the inputs are sorted?

File: src/dynamic_programming/subset_sum.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+/// # Arguments
+///
+/// * `nums` - A slice of unsigned 32-bit integers representing the input numbers.
+/// * `target_sum` - The target sum to which subsets are sought.
+///
+/// # Returns
+///
+/// An option containing a vector of vectors representing subsets that sum up to the target sum,
+/// or `None` if no such subsets exist.
+pub fn subset_sum(nums: &[u32], target_sum: u32) -> Option<Vec<Vec<u32>>> {
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:12]: It is still not clear to me what are you aiming for. If you want to handle a _general case_ (i.e. duplicates in `nums`), I would suggest to return a _set of set of indices_, for example for the input `[2, 3, 2]` and target `5`, the function should return `{{0, 1}, {0, 2}}`.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/dynamic_programming/subset_sum.rs (hunk 8 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:132]: The algorithm uses dynamic programming to generate subsets and does not assume any specific order of input elements.

File: src/dynamic_programming/subset_sum.rs (hunk 9 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:12]: Yeah, in some cases, when returned results for sample input `[2, 3, 2]` and target `5` are `[2, 3]` and `[3, 2]`. The code would only return `[2, 3]` (because `[3, 2]` is a duplication). Returning indices instead of numbers is a good idea. I will try it.

File: src/dynamic_programming/subset_sum.rs (hunk 7 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:15]: Since the table is used to store information about subsets that sum up to specific values, a more descriptive name could improve readability and maintainability of the code. I have changed the variable name from `dp` to `subset_table`.

Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:15]: I have merged the processes of initialization and populating `subset_table` into a single function `generate_subset_table`.

File: src/dynamic_programming/subset_sum.rs (hunk 9 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/subset_sum.rs:12]: I have updated the implementation to make `find_subset_with_sum` function (formerly `subset_sum`) return vector of vector representing indices of elements in the input `nums` vector that form subsets summing up to the target sum, or `None` if no such subsets exist.

File: src/dynamic_programming/subset_sum.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            )*
+        }
+    }
+
+    test_subset_sum! {
+        test_subset_sum_basic: (vec![5, 1, 3, 4, 2], 5, Some(vec![vec![0], vec![1, 3], vec![2, 4]])),
+        test_subset_sum_empty_input: (vec![], 10, None),
+        test_subset_sum_no_solution: (vec![1, 3, 2, 5, 4], 100, None),
+        test_subset_sum_single_element: (vec![7], 7, Some(vec![vec![0]])),
+        test_subset_sum_with_duplicates: (vec![1, 3, 2, 2], 5, Some(vec![vec![0, 2, 2], vec![1, 2]])),
```
Comment (vil02) [inline src/dynamic_programming/subset_sum.rs:125]:
```suggestion
        test_subset_sum_with_duplicates: (vec![1, 3, 2, 2], 5, Some(vec![vec![0, 2, 2], vec![1, 2]])),
```

There is something wrong with the implementation. You allow `vec![0, 2, 2]` but not `vec![0, 0, 0, 0, 0]`. I thin the goal was to use each entry of the input only once.

**Comment by TruongNhanNguyen** [comment]
I have try some methods to reconstruct the vectors of vectors (and sets of sets) representing indices of elements in `nums` array that sum up to specific value using dynamic programming, but it not yet successful. May be implementing the subset sum using dynamic programming isn't an effective way to generate all possible vectors of indices vector (because of time and space complexity grown fast when input size increase). Other effective way to do that is using backtracking strategy.

May this pull request will be closed soon. Thanks for your carefully review @vil02.
